tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp

```diff
@@ -2305,20 +2305,69 @@ bool BuildImmInit(Function *Ctor) {
   return true;
 }
 
+void CallCtorFunctionsAtInsertPt(llvm::Module &M,
+                                 llvm::SmallVector<llvm::Function *, 2> &Ctors,
+                                 Instruction *InsertPt) {
+  IRBuilder<> Builder(InsertPt);
+  for (Function *Ctor : Ctors) {
+    Builder.CreateCall(Ctor);
+  }
+}
+
+// If static initializers contain calls to external functions, this can
+// introduce inter-module init function ordering dependencies.  Some
+// dependencies may even introduce contradictions.  Creating and implementing an
+// intuitive standard approach to solve this is likely quite difficult.  Better
+// to disallow the ambiguous and unlikely case for now.
+bool IsValidCtorFunction(Function *F) {
+  for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+    if (CallInst *CI = dyn_cast<CallInst>(&(*I))) {
+      Function *Callee = CI->getCalledFunction();
+      if (Callee->isIntrinsic())
+        continue;
+      if (!Callee->isDeclaration())
+        continue;
+      if (hlsl::GetHLOpcodeGroup(Callee) != hlsl::HLOpcodeGroup::NotHL)
```

Comment:
Are there potentially a few special functions that we add during codegen that are not HL intrinsics or user functions?  Or do these all have another HLOpcodeGroup to identify them with?